src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Resources/MarkAssembliesWithNeutralResourcesLanguage.cs

```diff
@@ -68,28 +68,48 @@ public override void Initialize(AnalysisContext context)
                     return;
                 }
 
-                var hasResource = false;
-                RegisterAttributeAnalyzer(context, () => hasResource = true, generatedCode);
+                var alreadyReportedDiagnostic = new StrongBox<bool>(false);
+                var @lock = new object();
 
-                context.RegisterCompilationEndAction(context =>
-                {
-                    // there is nothing to do.
-                    if (!hasResource)
+                RegisterAttributeAnalyzer(
+                    context,
+                    shouldAnalyze: () =>
                     {
-                        return;
-                    }
-
-                    if (data != null &&
-                        data.ApplicationSyntaxReference?.GetSyntax(context.CancellationToken) is { } attributeSyntax)
+                        // This value can only change from false to true. So no need to lock if it's already true here.
+                        if (alreadyReportedDiagnostic.Value)
+                        {
+                            return false;
+                        }
+
+                        lock (@lock)
+                        {
+                            return !alreadyReportedDiagnostic.Value;
+                        }
+                    },
+                    onResourceFound: context =>
                     {
-                        // we have the attribute but its doing it wrong.
-                        context.ReportDiagnostic(attributeSyntax.CreateDiagnostic(Rule));
-                        return;
-                    }
-
-                    // attribute just don't exist
-                    context.ReportNoLocationDiagnostic(Rule);
-                });
+                        lock (@lock)
+                        {
+                            if (alreadyReportedDiagnostic.Value)
+                            {
+                                return;
+                            }
+
+                            alreadyReportedDiagnostic.Value = true;
+
+                            if (data != null &&
+                                data.ApplicationSyntaxReference?.GetSyntax(context.CancellationToken) is { } attributeSyntax)
+                            {
+                                // we have the attribute but its doing it wrong.
+                                context.ReportDiagnostic(attributeSyntax.CreateDiagnostic(Rule));
```

Comment:
Diagnostic with location will always be reported if we find a source assembly attribute here, otherwise a no-location diagnostic will always be reported - I believe the logic here is deterministic because `data` is computed at CompilationStart and does not alter during subsequent callbacks.